packages/api/apollo-server/types.ts

```diff
@@ -1,13 +1,82 @@
 import { gql } from 'apollo-server-core'
 
 const Types = gql`
+  input UserInput {
+    id: String
+    firstName: String
+    lastName: String
+    utorID: String
+    userType: UserType
+  }
+
   type Query {
-    getUser: User
-    getUsers: [User]
+    getUser(id: String, utorID: String, email: String): User
+    getUsers(users: [UserInput!]!): [User!]!
+    getAssignment(id: String): Assignment
+    getAssignments: [Assignment!]!
+    getSubmission(id: String): Submission
+    getSubmissions(userID: String): [Submission!]!
+    getFiles(submissionID: String): [File!]!
+    getAnnotations(fileID: String): [Annotation!]!
   }
 
   type Mutation {
-    editUser(firstName: String, lastName: String): User
+    addUser(
+      id: String!
+      firstName: String
+      lastName: String
+      utorID: String!
+      email: String
+      userType: UserType
```

Comment:
We should probably separate the creation of each usertype to different mutations addStudent, addTA, addAdministrator.

Reply:
> We should probably separate the creation of each usertype to different mutations addStudent, addTA, addAdministrator.
> 
> Same with edit.

Is it because TA and instructor may not have a utorid?

Reply:
Yes, and there may be several differences between how you create users.  For instance, we may need to consider privileges. We could limit this route to only work for instructors, but that may limit design space in the future.

Reply:
> Yes, and there may be several differences between how you create users. For instance, we may need to consider privileges. We could limit this route to only work for instructors, but that may limit design space in the future.

Is it required to separate `User` type to `Student`, `TA` and `Instructor` as well? 

Will this change make the app unnecessarily complicated?

Reply:
You don't need separate the User type. It's only creation/update process that can be extremely different. In other cases, we will simply return a 403/404 if the user does not have access to it.

Just look at updateUser. Instructor can update themselves, TAs, and Students. TAs can only update themselves and Students, Students can only update themselves. This is all just the privilege check. 

I think separating the logic can actually simplify it.

Reply:
> You don't need separate the User type. It's only the CRUD of this entity that can be extremely different. In other cases, we will simply return a 403/404 if the user does not have access to it.
> 
> Just look at updateUser. Instructor can update themselves, TAs, and Students. TAs can only update themselves and Students, Students can only update themselves. This is all just the privilege check.
> 
> I think separating the logic can actually simplify it.

I was trying to make this change, but I didn't find significant difference between Student mutation vs TA/Instructor mutation. What makes them different other than utorID?

Reply:
The graphql part would look almost identical

Reply:
> The graphql part would look almost identical

Shall we only make `editUser` for each user type? Because create and delete operations can only be performed by insturcotors